Feature/32044 color variables api for plugins

bomberstudios opened the pull request:
For https://github.com/sketch-hq/Sketch/issues/32044

bomberstudios committed:
First passing test

bomberstudios committed:
Two tests working

bomberstudios committed:
Third test passing

bomberstudios committed:
All existing tests passing

bomberstudios committed:
document.swatches, work in progress

bomberstudios committed:
Expose Swatch object so it can be used to create swatches

bomberstudios committed:
Swatches should always have a name

bomberstudios committed:
Document.swatches get/set API, work it progress

bomberstudios committed:
Add tests for remove methods

bomberstudios committed:
Added MSSwatch to .eslintrc

bomberstudios committed:
Added placeholder globalSwatchesDescriptor to make lint happy

bomberstudios committed:
Turns out we're not exposing swatches globally, since they'll live inside a Library

bomberstudios committed:
eslint is happy now

bomberstudios committed:
Fix failing test

christianklotz commented on Source/dom/models/Document.js:
```diff
@@ -686,6 +687,59 @@ Document.define('gradients', {
+        assets.addSwatchWithName_color(c.name, color)
+      })
+  },
+  insertItem(swatch, index) {
+    if (this.isImmutable()) {
+      return
+    }
+    const assets = this._getMSDocumentData().sharedSwatches()
+    const wrapped = Swatch.from(swatch)
+    const color = Color.from(wrapped.color).toMSColor()
+    if (index != undefined) {
+      // TODO: do something to insert the swatch at the right position
+      assets.addSwatchWithName_color_(wrapped.name, color)
+    } else {
+      assets.addSwatchWithName_color_(wrapped.name, color)
+    }
+  },
+  removeItem(index){
+    if (this.isImmutable()) {
+      return
+    }
+    const assets = this._getMSDocumentData().sharedSwatches()
+    const swatch = assets.swatchAt(index)
+    assets.removeSharedObjects([swatch])
+  },
```
If swatches are internally an unordered sequence it may be better to implement it as a set instead of an array 🤔

bomberstudios replied:
AFAIK they are ordered (or so I wrote on my dev notes: "Internally, swatches are returned in order, and apparently the order does not change when changing a Swatch name")

bomberstudios committed:
Cleanup

bomberstudios committed:
Fix lint warning

bomberstudios committed:
Swatches can now be imported

christianklotz committed:
Enable document logging test

The test failed because `swatches` was defined as `colors` on the immutable object

bomberstudios committed:
Fixes lint errors

christianklotz committed:
Delete comment
